server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -430,11 +431,15 @@ public void applyClusterState(ClusterChangedEvent event) {
     private boolean assertConsistentWithClusterState(ClusterState state) {
         final SnapshotsInProgress snapshotsInProgress = state.custom(SnapshotsInProgress.TYPE);
         if (snapshotsInProgress != null && snapshotsInProgress.entries().isEmpty() == false) {
-            final Set<Snapshot> runningSnapshots =
-                snapshotsInProgress.entries().stream().map(SnapshotsInProgress.Entry::snapshot).collect(Collectors.toSet());
-            final Set<Snapshot> snapshotListenerKeys = snapshotCompletionListeners.keySet();
-            assert runningSnapshots.containsAll(snapshotListenerKeys) : "Saw completion listeners for unknown snapshots in "
-                + snapshotListenerKeys + " but running snapshots are " + runningSnapshots;
+            synchronized (endingSnapshots) {
```

Comment:
The synchronization here isn't 100% necessary, could as well just copy the map before checking the listener key-set, but I figured the logic is a little easier to follow if we prevent concurrently starting another snapshot finalization.